packages/workspace/src/tasks-runner/task-orderer.ts

```diff
@@ -18,26 +24,84 @@ export class TaskOrderer {
     )
       return [tasks];
     if (tasks.length === 0) return [];
-    const res = [];
+    const stages: Task[][] = [];
+    const taskGraph = this.createTaskGraph(tasks);
     this.topologicallySortTasks(tasks).forEach((t) => {
-      const stageWithNoDeps = res.find(
-        (tasksInStage) => !this.taskDependsOnDeps(t, tasksInStage)
-      );
-      if (stageWithNoDeps) {
-        stageWithNoDeps.push(t);
+      const earliestStage = this.findEarliestStage(t, stages, taskGraph);
+      if (earliestStage) {
+        earliestStage.push(t);
       } else {
-        res.push([t]);
+        stages.push([t]);
       }
     });
-    return res;
+    return stages;
+  }
+
+  private createTaskGraph(tasks: Task[]): TaskGraph {
+    const graph: TaskGraph = {
+      tasks: [],
+      dependencies: {},
+    };
+    for (const task of tasks) {
+      graph.tasks.push(task);
+      graph.dependencies[task.id] = [];
+      const dependencyConfigs = getDependencyConfigs(
+        task.target,
+        this.projectGraph
+      );
+      const projectDependencies = new Set(
+        this.projectGraph.dependencies[task.target.project].map(
+          (dependency) => dependency.target
+        )
+      );
+
+      if (!dependencyConfigs) {
+        for (const _task of tasks) {
```

Comment:
don't use _ in names. We don't have it as a convention, so it doesn't like the rest of the codebase